### ISSUE TITLE
CORS not working with Client4

### ISSUE THREAD

**Issue by iampeter** [mattermost/mattermost-redux#557]
#### Summary
CORS seem to fail with `Client4` although mattermost-server has `AllowCorsFrom` set to `"*"`

#### Environment Information
- Chrome 65
- Webapp
- Mattermost Server 5.0.0

#### Steps to reproduce
Try to call `Client4.getMe()`

#### Expected behavior
Getting the results from the server.

#### Observed behavior
Chrome says: `Failed to load http://localhost:8065/api/v4/users/me: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Credentials' header in the response is '' which must be 'true' when the request's credentials mode is 'include'. Origin 'http://localhost:8081' is therefore not allowed access.`

A simple test posted here:
https://github.com/mattermost/mattermost-server/issues/8687
using `fetch` (`fetch('http://localhost:8065/api/v4/system/ping')`) works.

**Comment by lieut-data** [comment]
Oh, I see, it's because mattermost-redux passes `credentials: 'include'` to the `fetch` call that triggers the `CORS` check. I can reproduce the issue on the original test case if I do same.

**Comment by lieut-data** [comment]
It looks like the server is correct in /not/ sending `*`, since according to [discussions](https://stackoverflow.com/questions/12830095/setting-http-headers) and the wikipedia [article](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing) on this:
> The value of "*" is special in that it does not allow requests to supply credentials, meaning it does not allow HTTP authentication, client-side SSL certificates, or cookies to be sent in the cross-domain request.[8]

So it wouldn't work anyway. Given that the `Origin` header can only be set by the user-agent, it may simply not be possible to work around this for `file://` assets.

You'll probably need to front your local assets behind /something/ serving them up on localhost, e.g. as rudimentary as [described](http://php.net/manual/en/features.commandline.webserver.php):
```
> cd ~/files
> php -S localhost:8000
```

**Comment by iampeter** [comment]
@lieut-data I have changed the setting from `"*"` to `"localhost:8081"` (calling app), and now I'm getting:

```
Failed to load http://localhost:8065/api/v4/users/me: Response to preflight request doesn't pass access control check: 
No 'Access-Control-Allow-Origin' header is present on the requested resource. 
Origin 'http://localhost:8081' is therefore not allowed access. 
If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

**Comment by lieut-data** [comment]
@iampeter, hmm. With the setting changed back to `*`, can you capture the headers emitted in the `OPTIONS` request from your browser? I'm specifically interested in the value of the `Origin` header emitted by your user agent.

**Comment by iampeter** [comment]
@lieut-data here you go:

```
OPTIONS /api/v4/users/me HTTP/1.1
Host: localhost:8065
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
Access-Control-Request-Method: GET
Origin: http://localhost:8081
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.162 Safari/537.36
Access-Control-Request-Headers: authorization,x-requested-with
Accept: */*
Referer: http://localhost:8081/suggest
Accept-Encoding: gzip, deflate, br
Accept-Language: pl-PL,pl;q=0.9,en-US;q=0.8,en;q=0.7
```

**Comment by lieut-data** [comment]
Thanks for your patience, @iampeter -- indeed, the server is omitting the required `Access-Control-Allow-Credentials` response in this case. It looks like I got lost in unrelated details to miss the key part of your original report ;P

We'll get this one fixed!

**Comment by lieut-data** [comment]
I'm tracking the requisite server investigation and fix [here](https://mattermost.atlassian.net/browse/MM-11160).

**Comment by iampeter** [comment]
@lieut-data thanks, and I am also wondering how soon do you plan to take to this, because it is sort of urgent for  me :)

**Comment by lieut-data** [comment]
I'd imagine the soonest we can evaluate a fix for this would be Mattermost 5.2, scheduled for August 16th.

Are you running against a stock Mattermost install or a development instance? It's easy enough to make the change locally if you're iterating on development -- let me know if I can help with that. I've marked this as a security ticket (not in a bug sense but a feature sense), and hope to have the team make a decision on the best default behaviour for all customers going forward.

**Comment by iampeter** [comment]
@lieut-data I'm running a stock one, but can switch to a custom build.

**Comment by iampeter** [comment]
@lieut-data is there a proper way to `make build` a team edition without docker? 

getting lots of errors.

**Comment by iampeter** [comment]
@lieut-data ok building `release-5.0` worked fine for me, it was only master that caused errors

**Comment by lieut-data** [comment]
@iampeter, cool! Do you need help making the specific changes, or have you already figured that part out?

**Comment by iampeter** [comment]
@lieut-data thanks, I added it myself.

**Comment by njifares** [comment]
I'm facing the same issue: I'm trying to log in via an angular application (localhost:4200) against a Mattermost Server that is deployed on Kubernetes. 

The following code results in a HTTP Error (see below)

`try {
      Client4.setUrl(SITE_URL);
      await Client4.login(USERNAME, PASSWORD);
      const user = await Client4.getMe();
      console.log('user', user);
    } catch (error) {
      console.error(error);
    }
`

`Access to fetch at 'https://[[SITE_URL]]/api/v4/users/login' from origin 'http://localhost:4200' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.`

I've enable cross-origin requests from `localhost:4200`, `http://localhost:4200`, `http://localhost`, `localhost` and `*` but without any luck. 

I'm using 
- Mattermost Version: 5.13.0
- Database Schema Version: 5.13.0
- Database: postgres
- Angular: 8.2 (no redux)
- mattermost-redux: 5.14.0

Am I missing something in the configuration or set up?

**Comment by lieut-data** [comment]
@njifares, I'm not sure specifically what might be going on. Would you consider posting these details on https://community.mattermost.com/core/channels/peer-to-peer-help to see if someone can help root out the issue in your case?

**Comment by njifares** [comment]
I wasn't sure whether to create a new issue but since it's the same error I'm getting I'm posting my findings here.

#### Summary
Login preflight call fails using the Client4 (without redux) login method due to the origin header not being set. 

#### Environment Information
- Everything runs localhost
- Webapp or React Native app: Angular 8.2
- Mattermost Server Version:  5.14.0 (using docker container mattermost-preview)
- Database Schema Version: 5.14.0
- Database: mysql

#### Steps to reproduce
1) set up the docker image `mattermost-preview`
2) Set the siteUrl to localhost:{{DockerPort}}
3) Enable cross-origin requests from: * (for the sake of development)
4) Install mattermost-redux in an ng8 project 
5) Run following code
`async loginAndGetUserAync() {
    try {
      Client4.setUrl('http://localhost:{{DockerPort}}');
      await Client4.login({{Username}}, {{Pasword}});
      const user = await Client4.getMe();
      console.log('user', user);
    } catch (error) {
      console.error(error);
    }
  }`

#### Expected behavior
I expect the login call to succeed

#### Observed behavior
`Access to fetch at 'https://[[SITE_URL]]/api/v4/users/login' from origin 'http://localhost:4200' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.`

![Screenshot 2019-09-09 at 10 42 20 AM](https://user-images.githubusercontent.com/6762294/64558905-4d06b480-d345-11e9-998a-858812997460.png)


#### Possible fixes
The reason why it's failing is because the `origin` header is not set and therefore the preflight call is being blocked by the server. A solution would be to make it possible to add/set custom headers

**Comment by Willyfrog** [comment]
just a minor note, origin is not for preflight but for direct cors calls, which is the one happening on @njifares case

**Comment by njifares** [comment]
Setting the header via `setOptions` seems to do the job.

![Screenshot 2019-09-10 at 9 26 33 AM](https://user-images.githubusercontent.com/6762294/64593951-59bff280-d3af-11e9-923f-592795488951.png)

**Comment by huy-lv** [comment]
> Setting the header via `getOptions` seems to do the job.
> 
> ![Screenshot 2019-09-10 at 9 26 33 AM](https://user-images.githubusercontent.com/6762294/64593951-59bff280-d3af-11e9-923f-592795488951.png)

I have called setUrl() and getOptions() function like you but it still get the same problem :(

**Comment by ivan-kleshnin** [comment]
> I have called setUrl() and getOptions() function like you but it still get the same problem :(

`getOptions` does not set headers, it just returns a new object https://github.com/mattermost/mattermost-redux/blob/master/src/client/client4.ts#L440

It seems the only way to customize request headers in `Client4` is to extend a class, unfortunately.

---

To fix CORS errors with cloud-based MM API you can try:

#### Step-1

Use a reverse-proxy: https://github.com/Rob--W/cors-anywhere

#### Step-2

Extend the original `Client4` class and use token-based authentication

```tsx
import BaseClient4 from "mattermost-redux/client/client4"

class Client4 extends BaseClient4 {
  includeCookies = false // default `true` value won't work with reverse proxy
  defaultHeaders = {
    "Content-Type": "application/json", // fix defaults
  }

  // Original `login` does not return response headers so we have to make our own to get `resp.headers.token`!
  // Below is a simplified version of `login` with no fuss
  // https://github.com/mattermost/mattermost-redux/blob/master/src/client/client4.ts#L678
  login2 = async (username : string, password : string) : UserProfile & {token : string} => {
    this.trackEvent("api", "api_users_login")

    const resp = await this.doFetchWithResponse<UserProfile>(`${this.getUsersRoute()}/login`, {
      method: "POST",
      body: JSON.stringify({
        login_id: username,
        password,
      })
    })

    return {...resp.data, token: resp.headers.get("Token") || ""}
  }
}

const client4 = new Client4()
client4.setUrl("https://proxy-url/https://organization-name.cloud.mattermost.com") // e.g.

async function loginAndGetUser(username : string, password : string) {
  const userProfile = await client4.login2(username, password)
  console.log("userProfile:", userProfile)

  client4.setToken(userProfile.token)

  const me = await client4.getMe() // just to demonstrate – yields the same props as the above
  console.log("me:", me)
}

// ...
await loginAndGetUser(...) 
// ...
```
